flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/proxy/KinesisProxy.java

```diff
@@ -0,0 +1,211 @@
+						regionId,
+						stream,
+						shard.getShardId(),
+						shard.getSequenceNumberRange().getStartingSequenceNumber(),
+						shard.getSequenceNumberRange().getEndingSequenceNumber(),
+						shard.getParentShardId(),
+						shard.getAdjacentParentShardId()));
+				}
+				lastSeenShardId = shards.get(shards.size() - 1).getShardId();
+			} while (describeStreamResult.getStreamDescription().isHasMoreShards());
+		}
+		return shardList;
+	}
+
+	/**
+	 * Get a shard iterator for a Kinesis shard
+	 *
+	 * @param shard the shard to get the iterator for
+	 * @param shardIteratorType the iterator type to get
+	 * @param startingSeqNum the sequence number that the iterator will start from
+	 * @return the shard iterator
+	 */
+	public String getShardIterator(KinesisStreamShard shard, String shardIteratorType, String startingSeqNum) {
+		return kinesisClient.getShardIterator(shard.getStreamName(), shard.getShardId(), shardIteratorType, startingSeqNum).getShardIterator();
+	}
+
+	/**
+	 * Get metainfo for a Kinesis stream, which contains information about which shards this Kinesis stream possess.
+	 *
+	 * @param streamName the stream to describe
+	 * @param startShardId which shard to start with for this describe operation (earlier shard's infos will not appear in result)
+	 * @return the result of the describe stream operation
+	 */
+	private DescribeStreamResult describeStream(String streamName, String startShardId) {
+		final DescribeStreamRequest describeStreamRequest = new DescribeStreamRequest();
+		describeStreamRequest.setStreamName(streamName);
+		describeStreamRequest.setExclusiveStartShardId(startShardId);
+
+		DescribeStreamResult describeStreamResult = null;
+		String streamStatus = null;
+		int remainingRetryTimes = Integer.valueOf(
+			configProps.getProperty(KinesisConfigConstants.CONFIG_STREAM_DESCRIBE_RETRIES, Integer.toString(KinesisConfigConstants.DEFAULT_STREAM_DESCRIBE_RETRY_TIMES)));
+		long describeStreamBackoffTimeInMillis = Long.valueOf(
+			configProps.getProperty(KinesisConfigConstants.CONFIG_STREAM_DESCRIBE_BACKOFF, Long.toString(KinesisConfigConstants.DEFAULT_STREAM_DESCRIBE_BACKOFF)));
+
+		// Call DescribeStream, with backoff and retries (if we get LimitExceededException).
+		while ((remainingRetryTimes >= 0) && (describeStreamResult == null)) {
+			try {
+				describeStreamResult = kinesisClient.describeStream(describeStreamRequest);
+				streamStatus = describeStreamResult.getStreamDescription().getStreamStatus();
+			} catch (LimitExceededException le) {
+				LOG.warn("Got LimitExceededException when describing stream " + streamName + ". Backing off for "
+					+ describeStreamBackoffTimeInMillis + " millis.");
+				try {
+					Thread.sleep(describeStreamBackoffTimeInMillis);
+				} catch (InterruptedException ie) {
+					LOG.debug("Stream " + streamName + " : Sleep  was interrupted ", ie);
+				}
+			} catch (ResourceNotFoundException re) {
+				throw new RuntimeException("No stream name");
```

Comment:
I would change the message to something more generic + forward the exception, because it contains helpful details:
```
throw new RuntimeException("Error while getting stream details", re);
```